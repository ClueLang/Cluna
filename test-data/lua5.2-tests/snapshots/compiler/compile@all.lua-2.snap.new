---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/all.lua
---
local version = "Lua 5.2"
if _VERSION ~= version {
    io.stderr::write("\nThis test suite is for ", version, ", not for ", _VERSION, "\nExiting tests\n");
    return 
}
_soft = false
_port = false
_no32 = false
_nomsg = false
_noposix = false
_nolonglong = false
_noformatA = false
local usertests = rawget(_G, "_U")
if usertests {
    _soft = true
    _port = true
    _no32 = true
    _nomsg = true
    _noposix = true
    _nolonglong = true
    _noformatA = true
}
if usertests {
    T = nil
}
T = rawget(_G, "T")
package.path = "?;./?.lua" .. package.path
math.randomseed(0);
collectgarbage("setstepmul", 200);
collectgarbage("setpause", 200);
print("current path:\n****" .. package.path .. "****\n");
local c = os.clock()
local collectgarbage = collectgarbage
{
    local msgs = {}
    global fn Message(m) {
        if !_nomsg {
            print(m);
            msgs[(#msgs + 1)] = string.sub(m, 3,  - 3)
        }
    }
    assert(os.setlocale("C"));
    local T, print, format, write, assert, type, unpack, floor = T, print, string.format, io.write, assert, type, table.unpack, math.floor
    local fn F(m) {
        local fn round(m) {
            m = m + 0.04999
            return m - (m % 0.1)
        }
        if m < 1000 {
            return m
        } else {
            m = m / 1000
            if m < 1000 {
                return round(m) .. "K"
            } else {
                return round(m / 1000) .. "M"
            }
        }
    }
    local showmem
    if !T {
        local max = 0
        showmem = fn () {
            local m = collectgarbage("count") * 1024
            max = (m > max) && m || max
            print(format("    ---- total memory: %s, max memory: %s ----\n", F(m), F(max)));
        }
    } else {
        showmem = fn () {
            T.checkmemory();
            local total, numblocks, maxmem = T.totalmem()
            local count = collectgarbage("count")
            print(format("\n    ---- total memory: %s (%.0fK), max use: %s,  blocks: %d\n", F(total), count, F(maxmem), numblocks));
            print(format("\t(strings:  %d, tables: %d, functions: %d, " .. "\n\tudata: %d, threads: %d)", T.totalmem("string"), T.totalmem("table"), T.totalmem("function"), T.totalmem("userdata"), T.totalmem("thread")));
        }
    }
    local fn report(n) {
        print("\n***** FILE '" .. n .. "'*****");
    }
    local olddofile = dofile
    dofile = fn (n) {
        showmem();
        report(n);
        local f = assert(loadfile(n))
        local b = string.dump(f)
        f = assert(load(b))
        return f()
    }
    dofile('main.lua');
    {
        local next, setmetatable, stderr = next, setmetatable, io.stderr
        local mt = {}
        mt.__gc = fn (o) {
            stderr::write('.');
            local n = setmetatable({}, mt)
            o = nil
            local a, b, c, d, e = nil
        }
        local n = setmetatable({}, mt)
    }
    report("gc.lua");
    local f = assert(loadfile('gc.lua'))
    f();
    collectgarbage("generational");
    dofile('db.lua');
    assert(dofile('calls.lua') == deep && deep);
    olddofile('strings.lua');
    olddofile('literals.lua');
    assert(dofile('attrib.lua') == 27);
    collectgarbage("incremental");
    olddofile('strings.lua');
    olddofile('literals.lua');
    dofile('constructs.lua');
    dofile('api.lua');
    collectgarbage("generational");
    collectgarbage("setpause", 200);
    collectgarbage("setmajorinc", 500);
    assert(dofile('locals.lua') == 5);
    dofile('constructs.lua');
    dofile('code.lua');
    if !_G._soft {
        report('big.lua');
        local f = coroutine.wrap(assert(loadfile('big.lua')))
        assert(f() == 'b');
        assert(f() == 'a');
    }
    dofile('nextvar.lua');
    dofile('pm.lua');
    dofile('api.lua');
    assert(dofile('events.lua') == 12);
    dofile('vararg.lua');
    dofile('closure.lua');
    dofile('coroutine.lua');
    dofile('goto.lua');
    dofile('errors.lua');
    dofile('math.lua');
    dofile('sort.lua');
    dofile('bitwise.lua');
    assert(dofile('verybig.lua') == 10);
    collectgarbage();
    dofile('files.lua');
    if #msgs > 0 {
        print("\ntests not performed:");
        for i = 1, #msgs {
            print(msgs[(i)]);
        }
        print();
    }
    print("final OK !!!");
    local debug = require("debug")
    debug.sethook(fn (a) {
        assert(type(a) == 'string');
    }, "cr");
    _G.showmem = showmem
}
local _G, showmem, print, format, clock, assert, open = _G, showmem, print, string.format, os.clock, assert, io.open
local fname = T && "time-debug.txt" || "time.txt"
local lasttime
if !usertests {
    local f = io.open(fname)
    if f {
        lasttime = assert(tonumber(f::read('*a')))
        f::close();
    } else {
        lasttime = nil
    }
}
print('cleaning all!!!!');
for n with pairs(_G) {
    if !({
        ___Glob = 1, 
        tostring = 1
    })[(n)] {
        _G[(n)] = nil
    }
}
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
showmem();
local time = clock() - c
print(format("\n\ntotal time: %.2f\n", time));
if !usertests {
    lasttime = lasttime || time
    local diff = (time - lasttime) / time
    local tolerance = 0.05
    assert(diff < tolerance && diff >  - tolerance);
    assert(open(fname, "w"))::write(time)::close();
}
