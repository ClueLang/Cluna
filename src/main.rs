#![allow(non_snake_case)]
#![allow(non_upper_case_globals)]

macro_rules! check {
    ($tocheck: expr) => {
        match $tocheck {
            Ok(t) => t,
            Err(e) => return Err(e.to_string()),
        }
    };
}

macro_rules! arg {
    ($name: expr) => {
        unsafe { $name }
    };
}

// mod compiler;
mod compiler;
mod parser;
mod scanner;

use clap::Parser;
use compiler::compile_tokens;
use parser::*;
use scanner::*;
use std::{fs, fs::File, io::prelude::*, path::Path, time::Instant};

pub static mut ENV_TOKENS: bool = false;
pub static mut ENV_STRUCT: bool = false;
pub static mut ENV_OUTPUT: bool = false;
pub static mut ENV_DONTSAVE: bool = false;
pub static mut ENV_PATHISCODE: bool = false;
pub static mut ENV_NOMULTILINE: bool = false;

#[derive(Parser)]
#[clap(about, version, long_about = None)]
struct Cli {
    /// The path to the directory where the *.lua files are located.
    /// Every directory inside the given directory will be checked too.
    /// If the path points to a single *.lua file, only that file will be compiled.
    #[clap(required_unless_present = "license")]
    path: Option<String>,

    /// The name the output file will have
    #[clap(default_value = "main", value_name = "OUTPUT FILE NAME")]
    outputname: String,

    /// Print license information
    #[clap(short = 'L', long, display_order = 1000)]
    license: bool,

    /// Print list of detected tokens in compiled files
    #[clap(long)]
    tokens: bool,

    /// Print syntax structure of the tokens of the compiled files
    #[clap(long)]
    r#struct: bool,

    /// Print output Clue code in the console
    #[clap(short, long)]
    output: bool,

    /// Don't save compiled code
    #[clap(short = 'D', long)]
    dontsave: bool,

    /// Treat PATH not as a path but as clue code
    #[clap(short, long)]
    pathiscode: bool,

    /// Don't use multiline strings in the output
    #[clap(long)]
    nomultiline: bool,
}

fn compile_code(code: String, name: String, scope: usize) -> Result<String, String> {
    let time = Instant::now();
    let (tokens, _comments): (Vec<Token>, Vec<Comment>) = scan_code(code, name.clone())?;
    if arg!(ENV_TOKENS) {
        println!("Scanned tokens of file \"{}\":\n{:#?}", name, tokens);
    }
    let ctokens = parse_tokens(tokens, name.clone())?;
    if arg!(ENV_STRUCT) {
        println!("Parsed structure of file \"{}\":\n{:#?}", name, ctokens);
    }

    let code = compile_tokens(scope, ctokens);
    if arg!(ENV_OUTPUT) {
        println!("Compiled Lua code of file \"{}\":\n{}", name, code);
    }
    println!(
        "Compiled file \"{}\" in {} seconds!",
        name,
        time.elapsed().as_secs_f32()
    );
    Ok(code)
}

fn compile_file(path: &Path, name: String, scope: usize) -> Result<String, String> {
    let mut code: String = String::new();
    check!(check!(File::open(path)).read_to_string(&mut code));
    Ok(compile_code(code, name, scope)?)
}

fn compile_folder(path: &Path, rpath: String) -> Result<(), String> {
    for entry in check!(fs::read_dir(path)) {
        let entry = check!(entry);
        let name: String = entry
            .path()
            .file_name()
            .unwrap()
            .to_string_lossy()
            .into_owned();
        let filePathName: String = format!("{}/{}", path.display(), name);
        let filepath: &Path = &Path::new(&filePathName);
        let rname = rpath.clone() + &name;
        if filepath.is_dir() {
            compile_folder(filepath, rname + ".")?;
        } else if filePathName.ends_with(".lua") {
            let code = compile_file(filepath, name, 2)?;
            let compiled_path = format!(
                "{}{}",
                filepath.display().to_string().strip_suffix(".lua").unwrap(),
                ".clue"
            );
            if !arg!(ENV_DONTSAVE) {
                check!(fs::write(compiled_path, code))
            }
        }
    }
    Ok(())
}

fn main() -> Result<(), String> {
    let cli = Cli::parse();
    if cli.license {
        println!("{}", include_str!("../LICENSE"));
        return Ok(());
    }
    unsafe {
        ENV_TOKENS = cli.tokens;
        ENV_STRUCT = cli.r#struct;
        ENV_OUTPUT = cli.output;
        ENV_DONTSAVE = cli.dontsave;
        ENV_PATHISCODE = cli.pathiscode;
        ENV_NOMULTILINE = cli.nomultiline;
    }
    let codepath = cli.path.unwrap();
    if arg!(ENV_PATHISCODE) {
        println!(
            "{}",
            compile_code(codepath.clone(), "(command line)".to_owned(), 0)?
        );
        return Ok(());
    }
    let path: &Path = Path::new(&codepath);

    if path.is_dir() {
        compile_folder(path, String::new())?;
    } else if path.is_file() {
        let compiled_name =
            String::from(path.display().to_string().strip_suffix(".lua").unwrap()) + ".clue";
        let code = compile_file(
            path,
            path.file_name().unwrap().to_string_lossy().into_owned(),
            0,
        )?;

        if !arg!(ENV_DONTSAVE) {
            check!(fs::write(compiled_name, code))
        }
    } else {
        return Err(String::from("The given path doesn't exist"));
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use crate::compile_folder;
    use std::path::Path;

    #[test]
    fn compile() {
        compile_folder(Path::new("tests/"), String::new()).unwrap();
    }
}
