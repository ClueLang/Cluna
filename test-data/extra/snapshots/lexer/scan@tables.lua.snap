---
source: src/main.rs
expression: scanned
input_file: test-data/extra/tables.lua
---
[
    Token {
        kind: Local,
        lexeme: Symbol(
            "local",
        ),
        position: Position {
            line: 1,
            column: 5,
            span: 0..5,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "myTable",
        ),
        position: Position {
            line: 1,
            column: 13,
            span: 6..13,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 1,
            column: 15,
            span: 14..15,
        },
    },
    Token {
        kind: LeftBrace,
        lexeme: Symbol(
            "{",
        ),
        position: Position {
            line: 1,
            column: 17,
            span: 16..17,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "key1",
        ),
        position: Position {
            line: 3,
            column: 9,
            span: 45..49,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 3,
            column: 11,
            span: 50..51,
        },
    },
    Token {
        kind: String,
        lexeme: Symbol(
            "\"value1\"",
        ),
        position: Position {
            line: 3,
            column: 20,
            span: 52..60,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 3,
            column: 21,
            span: 60..61,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "key2",
        ),
        position: Position {
            line: 4,
            column: 9,
            span: 66..70,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 4,
            column: 11,
            span: 71..72,
        },
    },
    Token {
        kind: Number,
        lexeme: Number(
            Decimal(
                Decimal {
                    before_decimal: "42",
                    after_decimal: "",
                },
            ),
        ),
        position: Position {
            line: 4,
            column: 14,
            span: 72..75,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 4,
            column: 15,
            span: 75..76,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "key3",
        ),
        position: Position {
            line: 5,
            column: 9,
            span: 81..85,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 5,
            column: 11,
            span: 86..87,
        },
    },
    Token {
        kind: True,
        lexeme: Symbol(
            "true",
        ),
        position: Position {
            line: 5,
            column: 16,
            span: 88..92,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 5,
            column: 17,
            span: 92..93,
        },
    },
    Token {
        kind: LeftBracket,
        lexeme: Symbol(
            "[",
        ),
        position: Position {
            line: 6,
            column: 6,
            span: 98..99,
        },
    },
    Token {
        kind: String,
        lexeme: Symbol(
            "\"key with spaces\"",
        ),
        position: Position {
            line: 6,
            column: 23,
            span: 99..116,
        },
    },
    Token {
        kind: RightBracket,
        lexeme: Symbol(
            "]",
        ),
        position: Position {
            line: 6,
            column: 24,
            span: 116..117,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 6,
            column: 26,
            span: 118..119,
        },
    },
    Token {
        kind: String,
        lexeme: Symbol(
            "\"value with spaces\"",
        ),
        position: Position {
            line: 6,
            column: 46,
            span: 120..139,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 6,
            column: 47,
            span: 139..140,
        },
    },
    Token {
        kind: LeftBracket,
        lexeme: Symbol(
            "[",
        ),
        position: Position {
            line: 7,
            column: 6,
            span: 145..146,
        },
    },
    Token {
        kind: Number,
        lexeme: Number(
            Decimal(
                Decimal {
                    before_decimal: "10",
                    after_decimal: "",
                },
            ),
        ),
        position: Position {
            line: 7,
            column: 8,
            span: 145..148,
        },
    },
    Token {
        kind: RightBracket,
        lexeme: Symbol(
            "]",
        ),
        position: Position {
            line: 7,
            column: 9,
            span: 148..149,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 7,
            column: 11,
            span: 150..151,
        },
    },
    Token {
        kind: String,
        lexeme: Symbol(
            "\"value with numeric key\"",
        ),
        position: Position {
            line: 7,
            column: 36,
            span: 152..176,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 7,
            column: 37,
            span: 176..177,
        },
    },
    Token {
        kind: LeftBracket,
        lexeme: Symbol(
            "[",
        ),
        position: Position {
            line: 8,
            column: 6,
            span: 182..183,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "a",
        ),
        position: Position {
            line: 8,
            column: 7,
            span: 183..184,
        },
    },
    Token {
        kind: RightBracket,
        lexeme: Symbol(
            "]",
        ),
        position: Position {
            line: 8,
            column: 8,
            span: 184..185,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 8,
            column: 10,
            span: 186..187,
        },
    },
    Token {
        kind: String,
        lexeme: Symbol(
            "\"value with variable key\"",
        ),
        position: Position {
            line: 8,
            column: 36,
            span: 188..213,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 8,
            column: 37,
            span: 213..214,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "nestedTable",
        ),
        position: Position {
            line: 10,
            column: 16,
            span: 240..251,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 10,
            column: 18,
            span: 252..253,
        },
    },
    Token {
        kind: LeftBrace,
        lexeme: Symbol(
            "{",
        ),
        position: Position {
            line: 10,
            column: 20,
            span: 254..255,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "nestedKey",
        ),
        position: Position {
            line: 11,
            column: 18,
            span: 264..273,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 11,
            column: 20,
            span: 274..275,
        },
    },
    Token {
        kind: String,
        lexeme: Symbol(
            "\"nestedValue\"",
        ),
        position: Position {
            line: 11,
            column: 34,
            span: 276..289,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 11,
            column: 35,
            span: 289..290,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "anotherNestedTable",
        ),
        position: Position {
            line: 12,
            column: 27,
            span: 299..317,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 12,
            column: 29,
            span: 318..319,
        },
    },
    Token {
        kind: LeftBrace,
        lexeme: Symbol(
            "{",
        ),
        position: Position {
            line: 12,
            column: 31,
            span: 320..321,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "foo",
        ),
        position: Position {
            line: 13,
            column: 16,
            span: 334..337,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 13,
            column: 18,
            span: 338..339,
        },
    },
    Token {
        kind: String,
        lexeme: Symbol(
            "\"bar\"",
        ),
        position: Position {
            line: 13,
            column: 24,
            span: 340..345,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 13,
            column: 25,
            span: 345..346,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "hello",
        ),
        position: Position {
            line: 14,
            column: 18,
            span: 359..364,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 14,
            column: 20,
            span: 365..366,
        },
    },
    Token {
        kind: String,
        lexeme: Symbol(
            "\"world\"",
        ),
        position: Position {
            line: 14,
            column: 28,
            span: 367..374,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 14,
            column: 29,
            span: 374..375,
        },
    },
    Token {
        kind: RightBrace,
        lexeme: Symbol(
            "}",
        ),
        position: Position {
            line: 15,
            column: 10,
            span: 384..385,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 15,
            column: 11,
            span: 385..386,
        },
    },
    Token {
        kind: RightBrace,
        lexeme: Symbol(
            "}",
        ),
        position: Position {
            line: 16,
            column: 6,
            span: 391..392,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 16,
            column: 7,
            span: 392..393,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "array",
        ),
        position: Position {
            line: 19,
            column: 10,
            span: 413..418,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 19,
            column: 12,
            span: 419..420,
        },
    },
    Token {
        kind: LeftBrace,
        lexeme: Symbol(
            "{",
        ),
        position: Position {
            line: 19,
            column: 14,
            span: 421..422,
        },
    },
    Token {
        kind: Number,
        lexeme: Number(
            Decimal(
                Decimal {
                    before_decimal: "1",
                    after_decimal: "",
                },
            ),
        ),
        position: Position {
            line: 19,
            column: 16,
            span: 422..424,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 19,
            column: 17,
            span: 424..425,
        },
    },
    Token {
        kind: Number,
        lexeme: Number(
            Decimal(
                Decimal {
                    before_decimal: "2",
                    after_decimal: "",
                },
            ),
        ),
        position: Position {
            line: 19,
            column: 19,
            span: 425..427,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 19,
            column: 20,
            span: 427..428,
        },
    },
    Token {
        kind: Number,
        lexeme: Number(
            Decimal(
                Decimal {
                    before_decimal: "3",
                    after_decimal: "",
                },
            ),
        ),
        position: Position {
            line: 19,
            column: 22,
            span: 428..430,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 19,
            column: 23,
            span: 430..431,
        },
    },
    Token {
        kind: Number,
        lexeme: Number(
            Decimal(
                Decimal {
                    before_decimal: "4",
                    after_decimal: "",
                },
            ),
        ),
        position: Position {
            line: 19,
            column: 25,
            span: 431..433,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 19,
            column: 26,
            span: 433..434,
        },
    },
    Token {
        kind: Number,
        lexeme: Number(
            Decimal(
                Decimal {
                    before_decimal: "5",
                    after_decimal: "",
                },
            ),
        ),
        position: Position {
            line: 19,
            column: 28,
            span: 434..436,
        },
    },
    Token {
        kind: RightBrace,
        lexeme: Symbol(
            "}",
        ),
        position: Position {
            line: 19,
            column: 30,
            span: 437..438,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 19,
            column: 31,
            span: 438..439,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "mixedTable",
        ),
        position: Position {
            line: 22,
            column: 15,
            span: 470..480,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 22,
            column: 17,
            span: 481..482,
        },
    },
    Token {
        kind: LeftBrace,
        lexeme: Symbol(
            "{",
        ),
        position: Position {
            line: 22,
            column: 19,
            span: 483..484,
        },
    },
    Token {
        kind: String,
        lexeme: Symbol(
            "\"string value\"",
        ),
        position: Position {
            line: 23,
            column: 23,
            span: 493..507,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 23,
            column: 24,
            span: 507..508,
        },
    },
    Token {
        kind: Number,
        lexeme: Number(
            Decimal(
                Decimal {
                    before_decimal: "123",
                    after_decimal: "",
                },
            ),
        ),
        position: Position {
            line: 24,
            column: 12,
            span: 516..520,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 24,
            column: 13,
            span: 520..521,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "nested",
        ),
        position: Position {
            line: 25,
            column: 15,
            span: 530..536,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 25,
            column: 17,
            span: 537..538,
        },
    },
    Token {
        kind: LeftBrace,
        lexeme: Symbol(
            "{",
        ),
        position: Position {
            line: 25,
            column: 19,
            span: 539..540,
        },
    },
    Token {
        kind: True,
        lexeme: Symbol(
            "true",
        ),
        position: Position {
            line: 26,
            column: 17,
            span: 553..557,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 26,
            column: 18,
            span: 557..558,
        },
    },
    Token {
        kind: False,
        lexeme: Symbol(
            "false",
        ),
        position: Position {
            line: 27,
            column: 18,
            span: 571..576,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 27,
            column: 19,
            span: 576..577,
        },
    },
    Token {
        kind: RightBrace,
        lexeme: Symbol(
            "}",
        ),
        position: Position {
            line: 28,
            column: 10,
            span: 586..587,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 28,
            column: 11,
            span: 587..588,
        },
    },
    Token {
        kind: RightBrace,
        lexeme: Symbol(
            "}",
        ),
        position: Position {
            line: 29,
            column: 6,
            span: 593..594,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 29,
            column: 7,
            span: 594..595,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "key4",
        ),
        position: Position {
            line: 32,
            column: 9,
            span: 639..643,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 32,
            column: 11,
            span: 644..645,
        },
    },
    Token {
        kind: String,
        lexeme: Symbol(
            "\"value4\"",
        ),
        position: Position {
            line: 32,
            column: 20,
            span: 646..654,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 32,
            column: 21,
            span: 654..655,
        },
    },
    Token {
        kind: Identifier,
        lexeme: Symbol(
            "key5",
        ),
        position: Position {
            line: 33,
            column: 9,
            span: 660..664,
        },
    },
    Token {
        kind: Equals,
        lexeme: Symbol(
            "=",
        ),
        position: Position {
            line: 33,
            column: 11,
            span: 665..666,
        },
    },
    Token {
        kind: String,
        lexeme: Symbol(
            "\"value5\"",
        ),
        position: Position {
            line: 33,
            column: 20,
            span: 667..675,
        },
    },
    Token {
        kind: Comma,
        lexeme: Symbol(
            ",",
        ),
        position: Position {
            line: 33,
            column: 21,
            span: 675..676,
        },
    },
    Token {
        kind: RightBrace,
        lexeme: Symbol(
            "}",
        ),
        position: Position {
            line: 34,
            column: 2,
            span: 677..678,
        },
    },
]
